File: WarehouseApp/ViewModels/CajaViewModel.swift
import Foundation
import Combine

enum CajaEstado {
    case cerrada
    case abriendo
    case abierta(desde: Date)
    case cerrando
}

struct ResumenCaja {
    let apertura: Date
    let cierre: Date
    let totalVentas: Double
    let totalMP: Double
    let totalEfectivo: Double
    let cantidadVentas: Int
    let operadorName: String?
}

@MainActor
class CajaViewModel: ObservableObject {
    @Published var estado: CajaEstado = .cerrada
    @Published var totalVentas: Double = 0
    @Published var totalMP: Double = 0
    @Published var totalEfectivo: Double = 0
    @Published var cantidadVentas: Int = 0
    @Published var errorMessage: String?
    @Published var ultimoResumen: ResumenCaja?
    @Published private(set) var currentOperadorName: String?

    private let sessionKey = "caja_session_v2"

    init() { restoreEstado() }

    var estaAbierta: Bool {
        if case .abierta = estado { return true }
        return false
    }

    var horaApertura: Date? {
        if case .abierta(let d) = estado { return d }
        return nil
    }

    // MARK: - Abrir caja

    func abrirCaja(operador: Operador) async {
        guard case .cerrada = estado else { return }
        estado = .abriendo
        errorMessage = nil

        do {
            try await MercadoPagoService.verificarPOS()
        } catch {
            // Non-blocking: allow opening while offline
        }

        let ahora = Date()
        totalVentas = 0; totalMP = 0; totalEfectivo = 0; cantidadVentas = 0
        currentOperadorName = operador.name
        persistirSesion(apertura: ahora)
        estado = .abierta(desde: ahora)
    }

    // MARK: - Registrar venta

    func registrarVenta(total: Double, metodo: PaymentMethod) {
        guard total > 0 else { return }
        totalVentas += total
        cantidadVentas += 1
        switch metodo {
        case .qrMP, .pointMP: totalMP += total
        case .cash:            totalEfectivo += total
        }
        guard case .abierta(let apertura) = estado else { return }
        persistirSesion(apertura: apertura)
    }

    // MARK: - Cerrar caja

    func cerrarCaja() async {
        guard case .abierta(let apertura) = estado else { return }
        estado = .cerrando
        errorMessage = nil

        var totalMPConfirmado = totalMP
        do {
            totalMPConfirmado = try await MercadoPagoService.obtenerTotalPagos(desde: apertura)
        } catch {
            // Fallback to local total if API call fails
        }

        ultimoResumen = ResumenCaja(
            apertura: apertura,
            cierre: Date(),
            totalVentas: totalVentas,
            totalMP: totalMPConfirmado,
            totalEfectivo: totalEfectivo,
            cantidadVentas: cantidadVentas,
            operadorName: currentOperadorName
        )

        currentOperadorName = nil
        limpiarSesion()
        estado = .cerrada
    }

    // MARK: - Encrypted persistence

    private struct CajaSession: Codable {
        let apertura: Date
        var operadorName: String?
        let totalVentas: Double
        let totalMP: Double
        let totalEfectivo: Double
        let cantidadVentas: Int
    }

    private func persistirSesion(apertura: Date) {
        let session = CajaSession(
            apertura: apertura,
            operadorName: currentOperadorName,
            totalVentas: totalVentas,
            totalMP: totalMP,
            totalEfectivo: totalEfectivo,
            cantidadVentas: cantidadVentas
        )
        guard let encrypted = SecureStorage.encryptCodable(session) else { return }
        UserDefaults.standard.set(encrypted, forKey: sessionKey)
    }

    private func limpiarSesion() {
        UserDefaults.standard.removeObject(forKey: sessionKey)
        migrateLegacyKeys()
    }

    private func restoreEstado() {
        if let encrypted = UserDefaults.standard.data(forKey: sessionKey),
           let session: CajaSession = SecureStorage.decryptCodable(encrypted) {
            totalVentas         = session.totalVentas
            totalMP             = session.totalMP
            totalEfectivo       = session.totalEfectivo
            cantidadVentas      = session.cantidadVentas
            currentOperadorName = session.operadorName
            estado = .abierta(desde: session.apertura)
            return
        }
        // One-time migration from legacy plaintext keys
        let ud = UserDefaults.standard
        let legacyKey = "caja_apertura_date"
        if let apertura = ud.object(forKey: legacyKey) as? Date {
            totalVentas    = ud.double(forKey: "caja_total_ventas")
            totalMP        = ud.double(forKey: "caja_total_mp")
            totalEfectivo  = ud.double(forKey: "caja_total_efectivo")
            cantidadVentas = ud.integer(forKey: "caja_cantidad_ventas")
            estado = .abierta(desde: apertura)
            persistirSesion(apertura: apertura)
            migrateLegacyKeys()
        }
    }

    private func migrateLegacyKeys() {
        ["caja_apertura_date", "caja_total_ventas", "caja_total_mp",
         "caja_total_efectivo", "caja_cantidad_ventas"].forEach {
            UserDefaults.standard.removeObject(forKey: $0)
        }
    }
}

File: LektaAndroid/app/src/main/java/com/lekta/app/viewmodels/CajaViewModel.kt
package com.lekta.app.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.lekta.app.models.Operador
import com.lekta.app.models.PaymentMethod
import com.lekta.app.services.MercadoPagoService
import com.lekta.app.services.SecurePrefs
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import java.util.Date
import javax.inject.Inject

sealed class CajaEstado {
    data object Cerrada : CajaEstado()
    data object Abriendo : CajaEstado()
    data class Abierta(val desde: Long) : CajaEstado()
    data object Cerrando : CajaEstado()
}

data class ResumenCaja(
    val apertura: Long,
    val cierre: Long,
    val totalVentas: Double,
    val totalMP: Double,
    val totalEfectivo: Double,
    val cantidadVentas: Int,
    val operadorName: String?
)

@HiltViewModel
class CajaViewModel @Inject constructor(
    private val securePrefs: SecurePrefs,
    private val mpService: MercadoPagoService
) : ViewModel() {

    private val _estado = MutableStateFlow<CajaEstado>(CajaEstado.Cerrada)
    val estado: StateFlow<CajaEstado> = _estado

    private val _totalVentas = MutableStateFlow(0.0)
    val totalVentas: StateFlow<Double> = _totalVentas

    private val _totalMP = MutableStateFlow(0.0)
    val totalMP: StateFlow<Double> = _totalMP

    private val _totalEfectivo = MutableStateFlow(0.0)
    val totalEfectivo: StateFlow<Double> = _totalEfectivo

    private val _cantidadVentas = MutableStateFlow(0)
    val cantidadVentas: StateFlow<Int> = _cantidadVentas

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage

    private val _ultimoResumen = MutableStateFlow<ResumenCaja?>(null)
    val ultimoResumen: StateFlow<ResumenCaja?> = _ultimoResumen

    private val _currentOperadorName = MutableStateFlow<String?>(null)
    val currentOperadorName: StateFlow<String?> = _currentOperadorName

    private val sessionKey = "caja_session_v2"

    val estaAbierta: Boolean get() = _estado.value is CajaEstado.Abierta

    val horaApertura: Long?
        get() = (_estado.value as? CajaEstado.Abierta)?.desde

    init { restoreEstado() }

    fun abrirCaja(operador: Operador) {
        if (_estado.value !is CajaEstado.Cerrada) return
        _estado.value = CajaEstado.Abriendo
        _errorMessage.value = null

        viewModelScope.launch {
            try { mpService.verificarPOS() } catch (_: Exception) {}

            val ahora = System.currentTimeMillis()
            _totalVentas.value = 0.0
            _totalMP.value = 0.0
            _totalEfectivo.value = 0.0
            _cantidadVentas.value = 0
            _currentOperadorName.value = operador.name
            persistirSesion(ahora)
            _estado.value = CajaEstado.Abierta(ahora)
        }
    }

    fun registrarVenta(total: Double, metodo: PaymentMethod) {
        if (total <= 0) return
        _totalVentas.value += total
        _cantidadVentas.value += 1
        when (metodo) {
            PaymentMethod.QR_MP, PaymentMethod.POINT_MP -> _totalMP.value += total
            PaymentMethod.CASH -> _totalEfectivo.value += total
        }
        val apertura = ((_estado.value as? CajaEstado.Abierta)?.desde) ?: return
        persistirSesion(apertura)
    }

    fun cerrarCaja() {
        val apertura = (_estado.value as? CajaEstado.Abierta)?.desde ?: return
        _estado.value = CajaEstado.Cerrando
        _errorMessage.value = null

        viewModelScope.launch {
            var totalMPConfirmado = _totalMP.value
            try {
                totalMPConfirmado = mpService.obtenerTotalPagos(Date(apertura))
            } catch (_: Exception) {}

            _ultimoResumen.value = ResumenCaja(
                apertura = apertura,
                cierre = System.currentTimeMillis(),
                totalVentas = _totalVentas.value,
                totalMP = totalMPConfirmado,
                totalEfectivo = _totalEfectivo.value,
                cantidadVentas = _cantidadVentas.value,
                operadorName = _currentOperadorName.value
            )

            _currentOperadorName.value = null
            securePrefs.removeEncrypted(sessionKey)
            _estado.value = CajaEstado.Cerrada
        }
    }

    private data class CajaSession(
        val apertura: Long,
        val operadorName: String?,
        val totalVentas: Double,
        val totalMP: Double,
        val totalEfectivo: Double,
        val cantidadVentas: Int
    )

    private fun persistirSesion(apertura: Long) {
        securePrefs.saveEncrypted(sessionKey, CajaSession(
            apertura = apertura,
            operadorName = _currentOperadorName.value,
            totalVentas = _totalVentas.value,
            totalMP = _totalMP.value,
            totalEfectivo = _totalEfectivo.value,
            cantidadVentas = _cantidadVentas.value
        ))
    }

    private fun restoreEstado() {
        val session = securePrefs.loadEncrypted<CajaSession>(sessionKey) ?: return
        _totalVentas.value = session.totalVentas
        _totalMP.value = session.totalMP
        _totalEfectivo.value = session.totalEfectivo
        _cantidadVentas.value = session.cantidadVentas
        _currentOperadorName.value = session.operadorName
        _estado.value = CajaEstado.Abierta(session.apertura)
    }
}
